### PR TITLE
Add KeyManagedEncryptedEnv and AESBlockCipher

### PR DESCRIPTION
**Pull request by yiwu-arbug**
Summary:
Introduce `KeyManagedEncryptedEnv` which wraps around `EncryptedEnv` but provides an `KeyManager` API to enable key management per file. Also implements `AESBlockCipher` with OpenSSL.

Test Plan:
not tested yet. will update.

Signed-off-by: Yi Wu <yiwu@pingcap.com>

### CODE REVIEW

**Comment by zhangjinpeng1987** [comment]
Is it possible to implement it out of this repo?

**Comment by yiwu-arbug** [comment]
> Is it possible to implement it out of this repo, since it is a specific EncryptedEnv.

Since we haven't decided where the place the code, I'm just sending them here for preview.

File: include/rocksdb/encryption.h [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+  Status RenameFile(const std::string& src_fname,
+                    const std::string& dst_fname) override;
+
+ private:
+  const std::shared_ptr<KeyManager> key_manager_;
+  const std::unique_ptr<AESEncryptionProvider> provider_;
+  const std::unique_ptr<Env> encrypted_env_;
+};
+
+// TODO(yiwu): Return a status. If OpenSSL is not available at compile time,
```
[the remark this answers is not in the document]
Comment (yiwu-arbug) [reply include/rocksdb/encryption.h]: sure. Then I'll keep it as is.

**Pull request merged**